Library/Homebrew/cmd/install.rb

```diff
@@ -278,48 +278,57 @@ def install
     # formula was found, but there's a problem with its implementation).
     $stderr.puts e.backtrace if Homebrew::EnvConfig.developer?
     ofail e.message
-  rescue FormulaOrCaskUnavailableError => e
-    if e.name == "updog"
+  rescue FormulaOrCaskUnavailableError, Cask::CaskUnavailableError => e
+    # formula name or cask token
+    name = e.try(:name) || e.token
```

Comment:
I know that this is mostly existing code but: I wonder if we could punt pretty much the whole `rescue` block into a new method (perhaps in `search.rb`?) which can be called separately for formulae/casks. It feels like there's a lot of logic in here, more than you'd expect for exception handling. Would also allow this to be tested, too.

Given this isn't your mess: feel free to punt to a future PR, this is not blocking.